### ISSUE TITLE
Markdown support in documentation rendering of OMEdit

### ISSUE THREAD

**Issue by dietmarw** [OpenModelica/OpenModelica#9284]
### Description

It would be a nice extension if the documentation browser would support the rendering of markdown formatted text. This is something that has been added in the latest release of Dymola 2023, and although not standard would be nice to have. The LaTeX equation rendering is already working thanks to [#3311](https://trac.openmodelica.org/OpenModelica/ticket/3311).

**Comment by SG-phimeca** [comment]
I second this request. The documentation of ThermoSysPro (https://thermosyspro.gitlab.io/documentation/) has been written with Sphinx and the MyST parser plugin to support Markdown.
We plan to reintegrate the markdown source into the Modelica annotation for documenation so that Dymola users can access it directly from Dymola.
Enabling this in OpenModelica would be very useful to us.

**Comment by casella** [comment]
I agree this would be a nice feature. What about proposing if for the Modelica Specification? Then, we could do this as a second test implementation. @dietmarw what do you think?

**Comment by dietmarw** [comment]
The pragmatic approach would be to implement it now and also push this to tje specification. Otherwise OM will lack behind when it comes to Markdown support.

**Comment by casella** [comment]
Sure. @adeas31 what do you think the effort could be?

**Comment by Sonyoyo** [comment]
Any news on this issue? For ThermoSysPro we are planning to use Markdown for the documentation.

**Comment by casella** [comment]
@adeas31 ping! Maybe @arun3688 could also work on that?

**Comment by adeas31** [comment]
I will take a look at it. So instead of html we store markdown in the documentation info annotation, right?

**Comment by SG-phimeca** [comment]
Yes @adeas31 that's it. Is there any way we can help by rpvoding additional information?

**Comment by SG-phimeca** [comment]
Here we are used to both Pandoc (https://pandoc.org/MANUAL.html) and MyST (https://myst-parser.readthedocs.io/en/latest/intro.html) for compiling Markdown to Html and other formats. There may be some reusable component here?!

**Comment by casella** [comment]
I see that Pandoc is written in Haskell and MyST in Python. OMEdit is written in C++. I'm not sure we want to introduce dependencies onto other languages just because of this feature. @adeas31 what do you think?

**Comment by casella** [comment]
Maybe we could use https://github.com/progsource/maddy?

**Comment by bilderbuchi** [comment]
> I see that Pandoc is written in Haskell and MyST in Python. OMEdit is written in C++. I'm not sure we want to introduce dependencies onto other languages just because of this feature. @adeas31 what do you think?

At least Pandoc is a standalone tool, though, so I'm not sure why the implementation language is a concern? 

MyST of course needs a Python environment, but then you already have OMPython at least associated (not sure if default installed) with OMEdit, so it seems like that would also not be a big stepping stone?

Importantly, both Pandoc and MyST are well-established tools with a community and (at least in my perception) excellent track record. 
I don't know the thing you linked to, but from its [activity graph](https://github.com/progsource/maddy/graphs/contributors) it does not seem that maintained, and its bus factor is probably <=1.

**Comment by bilderbuchi** [comment]
If you end up adding markdown support to the specification, I recommend to specify which dialect/variant you mean (there are several), AFAIK, https://commonmark.org/ is a popular common denominator that is also well-specified and stable.

**Comment by SG-phimeca** [comment]
An important point is to support math rendering with LaTeX syntax.

**Comment by bilderbuchi** [comment]
> An important point is to support math rendering with LaTeX syntax.

Just to get this clarified: Do you mean precisely _LaTeX_ syntax, or do you mean (implicitly) "math typesetting" syntax (which MathML or AsciiMath could also deliver)? If the former, why that one, specifically?

**Comment by SG-phimeca** [comment]
I mean “as close as LaTeX as is reasonable”. MathJax Support is quite good. Both Pandoc and MyST rely on LateX (or a subset of) syntax.

**Comment by casella** [comment]
I don't want to sound pedantic, but we should rely on well-defined standards, otherwise we end up with something that only works in OpenModelica, which is really not our main purpose.

**Comment by SG-phimeca** [comment]
Completely agree. Especially, the markdown flavor should probably be the same or “compatible” with what is done in Dymola.
Developpers who strive to make their libraries compatible with both Dymola and OpenModelica should not have to choose for which tool they write the doc!

**Comment by casella** [comment]
BTW there are other Modelica tools around besides Dymola and OpenModelica, most notably Modelon Impact, Wolfram System Modeler, MapleSoft MapleSim,  ESI SimulationX, Siemens Amesim, etc.

**Comment by casella** [comment]
I just wonder, if there are so nice tools to translate markdown into HTML, shouldn't we rather find ways to integrate them in the workflow? For example, we could store the markdown source code _and_ the corresponding HTML code (which is standardized by the MLS) in the mode annotations. 

Markdown editing tools could be used to fetch the markdown source code from the Modelica code, edit it in a nice GUI, and then store the HTML code in the Modelica code again. This would ensure that all Modelica tools can display the documentation properly.

I think this would be a more scalable approach, following the Linux philosopy of combining many tools that each one thing very well, rather than by trying to integrate everything into a monolithic tool

What do you think?

**Comment by SG-phimeca** [comment]
The approach you suggest (relying on third party well established tool) could also reduce the documentation and maintenance effort.

Here are additional details about our perspective as users.
A few years ago, EDF and Phimeca decided to use the freshly published book on ThermoSysPro (https://thermosyspro.com/) by Baligh El Hefni and Daniel Bouskela to document the library.

We needed pictures, tables and equations. First, we disliked having to type HTML. Second, we did not manage to have the browsers included in the main target tools (Dymola and OpenModelica) to display equations without much tinkering.
We opted for a stand-alone wiki like doc (https://thermosyspro.gitlab.io/documentation/), produced with Sphinx and a plugin for typing the source in markdown (today we would use MyST/Jupyterbook, https://github.com/executablebooks/jupyter-book).

We discovered recently that Dymola support Markdown in Documentation annotations.
We would like to repatriate the documentation inside the Modelica code.
We have written a Python script to do so.
We are currently working on another script to automatically scrap the Markdown in a Modelica library Documentation annotations and create a MyST Jupyterbook.

**Comment by casella** [comment]
@SG-phimeca thanks for the clarification. The user requirement is pretty clear. On the other hand, introducing support for languages other than HTML in the MLS is a big deal, because a) is has to be clearly specified and b) all tools will be expected to provided it.

HTML in general is not meant to be written manually, people stopped writing HTML by hand more than 25 years ago, though I do this all the time for the documentation of my libraries 😄. It is a low-level language for content rendering, which has several big advantages:
- it is clearly specified by a non-profit entity
- it is well established
- it can render pretty much any content that can be displayed in a browser
- there are countless HTML rendering engines available (including open-source ones) for tool developers

So my question is: if you are automatically generating the documentation of your library via scripts, why using Markdown and not HTML?

Regarding the issue of representing equations, MathML has been around for a very long time, it is another well-established standard, a companion of HTML, supported by most HTML rendering engines. Again, MathML, as HTML, is not meant to be hand-written by humans, it is meant to be a standard language to encode mathematics in documentations. We could make it clearer in the MLS that HTML doumentation could also include MathML code and that should be supported.

**Comment by SG-phimeca** [comment]
Sorry for introducing confusion. I'll try to clarify.

Writing markdown is our way of easily generating html.

We want to write the documentation from within the tools (Dymola or OpenModelica), or possibly from text editor.
The source of the documentation should be in the models.
It would then be directly explorable within the tools.

Having the doc “externalized” as a stand-alone wiki (similar to Python module documentation with Sphinx) is a nice to have.

Reinjecting the markdown/html source from a stand-alone wiki to the Modelica code is a one shot script because we change our mind. Because appropriate support for equations and the possibility of typing the doc with a relatively natural syntax (markdown) appeared in Dymola, and could come into OpenModelica (this issue!).

Other users may offer alternate perspectives.

**Comment by Sonyoyo** [comment]
For information, since different markdown flavors exist, the documentation of the following model is properly displayed in recent versions of Dymola, including equations, tables and images (the file image.png is located in the package directory Test_image and referenced via a Modelica relative URl):

```
within Test_image;
model Test_markdown

  annotation (Documentation(info="# Markdown test model

Model to test Markdown documentation

## First points 

### Some maths

The *main* equation would be: \\\\(W_{TH}=\\alpha S(T_1-T_2)\\\\)

If you want to give it **more** importance (centered on its own line):

$$W_{TH}=hS(T_1-T_2)$$

### Some tables

Parameters | Description | Unit
---: | :--- | :---:
 \\\\(\\alpha\\\\) | Coefficient | \\\\(W/m^2K\\\\)
S | Exchange surface |\\\\(m^2\\\\)


## Some images

Using a *Modelica* relative path

![Test image](modelica://Test_image/image.png)

", revisions="First Trial
"));
end Test_markdown;
```

<img width="321" alt="image" src="https://github.com/user-attachments/assets/739ee3ea-cdec-4f9d-90ab-c3de2bf0bc5a" />

**Comment by casella** [comment]
What I don't like is that we start reverse-engineering Dymola features in OpenModelica. The OSMC supports the Modelica open standard.

I am fine with supporting markdown documentation in OpenModelica in addition to HTML, but only if this is standardized by the Modelica Association. 

Of course we don't necessarily need to wait until a new version of the specification is actually released in order to implement it, but there should be at least some kind of consensus in the MAP-Lang project that this is eventually going to end up in the standard. In that case, we are happy to provide test implementations.

@dietmarw,   @SG-phimeca, would you like to start the discussion on the [Modelica specification](https://github.com/modelica/ModelicaSpecification/issues) issue tracker?

**Comment by dietmarw** [comment]
@casella This has already been done 5 years ago https://github.com/modelica/ModelicaStandardLibrary/issues/3122 and obviously let to the pragmatic solution of Dymola to just implement it. Of course OpenModelica could wait another 5 or 10 years to wait for some decision OR they could just activate it in their tool to offer something useful to their users.

**Comment by casella** [comment]
@dietmarw I don't think adding documentation in a non standard format to the Modelica _Standard_ Library is the way to go. It should be exactly the other way round.

We need to have some discussion within MAP-Lang, and hear the opinon of all interested tool vendors. I'm not saying we need to wait another 5 or 10 years for that, but we need to have that discussion in the proper venue.

Do you agree?

**Comment by dietmarw** [comment]
In theory and an ideal world, sure. But the ticket I pointed to shows that it doesnt work that way. So I would not hold my breath if I was OM.

**Comment by casella** [comment]
We live in the best possible world 😃 

Let's see the reactions to modelica/ModelicaSpecification/issues#3636 and then we can take a motivated decision. Feel free to comment there.
